libs/langchain/langchain/schema/runnable/history.py

```diff
@@ -0,0 +1,254 @@
+        self, output_val: Union[str, BaseMessage, Sequence[BaseMessage], dict]
+    ) -> List[BaseMessage]:
+        from langchain.schema.messages import BaseMessage
+
+        if isinstance(output_val, dict):
+            output_val = output_val[self.output_messages_key or "output"]
+
+        if isinstance(output_val, str):
+            from langchain.schema.messages import AIMessage
+
+            return [AIMessage(content=output_val)]
+        elif isinstance(output_val, BaseMessage):
+            return [output_val]
+        elif isinstance(output_val, (list, tuple)):
+            return list(output_val)
+        else:
+            raise ValueError()
+
+    def _enter_history(self, input: Any, config: RunnableConfig) -> List[BaseMessage]:
+        hist: BaseChatMessageHistory = config["configurable"]["message_history"]
+        # return only historic messages
+        if self.message_history_key:
+            return hist.messages.copy()
+        # return all messages
+        else:
+            input_val = (
+                input if not self.input_messages_key else input[self.input_messages_key]
+            )
+            return hist.messages.copy() + self._get_input_messages(input_val)
+
+    async def _aenter_history(
+        self, input: Dict[str, Any], config: RunnableConfig
+    ) -> List[BaseMessage]:
+        return await asyncio.get_running_loop().run_in_executor(
+            None, self._enter_history, input, config
+        )
+
+    def _exit_history(self, run: Run, config: RunnableConfig) -> None:
+        hist: BaseChatMessageHistory = config["configurable"]["message_history"]
+
+        # Get the input messages
+        inputs = load(run.inputs)
+        input_val = inputs[self.input_messages_key or "input"]
+        input_messages = self._get_input_messages(input_val)
+
+        # Get the output messages
+        output_val = load(run.outputs)
+        output_messages = self._get_output_messages(output_val)
+
+        for m in input_messages + output_messages:
+            hist.add_message(m)
+
+    def _merge_configs(self, *configs: Optional[RunnableConfig]) -> RunnableConfig:
+        config = super()._merge_configs(*configs)
+        # extract thread_id
+        if "thread_id" not in config.get("configurable", {}):
+            example_input = {self.input_messages_key: "foo"}
+            example_config = {"configurable": {"thread_id": "123"}}
+            raise ValueError(
+                "thread_id is required."
+                " Pass it in as part of the config argument to .invoke() or .stream()"
+                f"\neg. chain.invoke({example_input}, {example_config})"
+            )
+        # attach message_history
+        thread_id = config["configurable"]["thread_id"]
+        user_id = config["configurable"].get("user_id")
+        session_id = thread_id if user_id is None else f"{user_id}:{thread_id}"
+        config["configurable"]["message_history"] = self.factory(
+            session_id,
```

Comment:
What if instead we let the factory accept a dict with inputs so we avoid doing the string concatenation here, then users won't have to create parsing code in the factory to determine if it was only thread id or user and thread id that were provided etc

```python
    def _merge_configs(self, *configs: Optional[RunnableConfig]) -> RunnableConfig:
        config = super()._merge_configs(*configs)
        # extract thread_id
        if "thread_id" not in config.get("configurable", {}):
            example_input = {self.input_messages_key: "foo"}
            example_config = {"configurable": {"thread_id": "123"}}
            raise ValueError(
                "thread_id is required."
                " Pass it in as part of the config argument to .invoke() or .stream()"
                f"\neg. chain.invoke({example_input}, {example_config})"
            )
        # attach message_history
        thread_id = config["configurable"]["thread_id"]
        user_id = config["configurable"].get("user_id")
        config["configurable"]["message_history"] = self.factory(
            {"thread_id": thread_id, "user_id": user_id}
        )
        return config
```

Reply:
no strong opinions on my end, thoughts @nfcampos?

Reply:
I didn't do that because none of our chat history classes accept such a dict, so all would need wrappers to be used here if we go with that

Reply:
What do we think if we reduce this to `session_id` as a default to match existing implementations? 

* It's possible to parse user id and thread id from the session id, so it won't block folks from implementing that.

---

In addition, we could do the following as an enhancement if we like it in the future:

* enhance functionality to accept a factory which takes either a dict or a str as the first argument and based on the type signature invoke it appropriately
* Allows user to override default config spec by passing something during initialization (and assume this only works with a factory that takes a dict)